### ISSUE TITLE
Protect ID Bug

### ISSUE THREAD

**Issue by joubertredrat** [joubertredrat/Matomo-ProtectTrackID#33]
Bug report received from e-mail.

> Hi,
>
> I'm at PHP version 8.0.9 and your latest version of Protect Track ID gives an error when Matomo build the reports, it puts this in matomo-archive.log en exits
> 
> Uncaught exception in /home/httpd/html/[dmp.imcreatives.com/public_html/matomo/plugins/ProtectTrackID/ProtectTrackID.php](http://dmp.imcreatives.com/public_html/matomo/plugins/ProtectTrackID/ProtectTrackID.php) line 101:
> syntax error, unexpected ')'
> 
> I have no idea what to do, I can't downgrade as Matomo requires this new version of your plugin
> 
> Thanks,
> 
> Marcel

**Comment by joubertredrat** [comment]
> Hi again ;-)
> 
> I got it fixed by replacing your new regex code to one from an older version… note that your codes is having a ) too much in it.. for some reason Matomo just worked fine with that extra ) but the reports build cron function not
> 
> So.. I replaced this...
> 
> ```php
> $regex = sprintf(
>             '/^(%s){%s}$/',
>             implode('|', str_split($settings->base)),
>             $settings->length,
>         );
>```
> 
> With…
> 
> `$regex = '/^('.implode('|', str_split($settings->base)).'){'.$settings->length.'}$/‘;`
> 
> And things are back working and reports being build again…
> 
> Cheers
> 
> Marcel

**Comment by joubertredrat** [comment]
> Hi,
> 
> Here you go
> 
> Matomo version:  4.13.1
> MySQL version:  5.5.68-MariaDB
> PHP version:  8.0.9
> 
> ProtectTrackID (v2.3.0) 
> 
> Cheers,
> 
> Marcel

**Comment by joubertredrat** [comment]
I tested here with same PHP version and wasn't possible to reproduce issue here. 

~~The strange behavior is about him log, the issue was in `ProtectTrackID.php` on line 101 about syntax error, but on version 2.3.0 this file have only 78 lines, as can be viewed  in this link https://github.com/joubertredrat/Matomo-ProtectTrackID/blob/2.3.0/ProtectTrackID.php#L78.~~ Mamoto injects code to be able to load plugin in Wordpress, because of this the difference about the line error.

This issue is about `isValidHash`, a function that validates if value received is hashed or not, then I'm going to investigate more about it.

```bash
dev@Devs-MacBook-Pro ~/source/joubertredrat/matomo (4.x-dev)> docker compose logs -f
matomo-matomo-1  | [Tue Jan 17 13:18:12 2023] PHP 8.0.9 Development Server (http://0.0.0.0:8080) started
matomo-matomo-1  | [Tue Jan 17 13:18:49 2023] 172.19.0.1:55352 Accepted
```

![screenshot-0 0 0 0_8080-2023 01 17-10_22_56](https://user-images.githubusercontent.com/1520407/212910548-d3121c43-ab9d-44a9-a243-812aff012788.png)
![screenshot-0 0 0 0_8080-2023 01 17-10_23_32](https://user-images.githubusercontent.com/1520407/212910589-34ed4d2c-95bc-403f-97c7-eb259983c707.png)
![screenshot-0 0 0 0_8080-2023 01 17-10_24_21](https://user-images.githubusercontent.com/1520407/212910611-1c3f1a4b-1b3f-473b-8b14-5b9b05580ac5.png)

**Comment by joubertredrat** [comment]
I executed console and was fine.
```bash
/var/www # ./console core:archive --url=http://127.0.0.1
INFO [2023-01-17 15:38:22] 55  ---------------------------
INFO [2023-01-17 15:38:22] 55  INIT
INFO [2023-01-17 15:38:22] 55  Running Matomo 4.13.1 as Super User
INFO [2023-01-17 15:38:22] 55  ---------------------------
INFO [2023-01-17 15:38:22] 55  NOTES
INFO [2023-01-17 15:38:22] 55  - If you execute this script at least once per hour (or more often) in a crontab, you may disable 'Browser trigger archiving' in Matomo UI > Settings > General Settings.
INFO [2023-01-17 15:38:22] 55    See the doc at: https://matomo.org/docs/setup-auto-archiving/
INFO [2023-01-17 15:38:22] 55  - Async process archiving supported, using CliMulti.
INFO [2023-01-17 15:38:22] 55  - Reports for today will be processed at most every 900 seconds. You can change this value in Matomo UI > Settings > General Settings.
INFO [2023-01-17 15:38:22] 55  ---------------------------
INFO [2023-01-17 15:38:22] 55  START
INFO [2023-01-17 15:38:22] 55  Starting Matomo reports archiving...
INFO [2023-01-17 15:38:22] 55  Start processing archives for site 1.
INFO [2023-01-17 15:38:22] 55    Will invalidate archived reports for today in site ID = 1's timezone (2023-01-17 00:00:00).
INFO [2023-01-17 15:38:25] 55  Archived website id 1, period = day, date = 2023-01-17, segment = '', 1 visits found. Time elapsed: 3.063s
INFO [2023-01-17 15:38:25] 55  Archived website id 1, period = day, date = 2022-12-26, segment = '', 2 visits found. Time elapsed: 3.066s
INFO [2023-01-17 15:38:25] 55  Archived website id 1, period = day, date = 2022-12-23, segment = '', 1 visits found. Time elapsed: 3.069s
INFO [2023-01-17 15:38:28] 55  Archived website id 1, period = week, date = 2023-01-16, segment = '', 1 visits found. Time elapsed: 3.064s
INFO [2023-01-17 15:38:28] 55  Archived website id 1, period = week, date = 2022-12-26, segment = '', 2 visits found. Time elapsed: 3.068s
INFO [2023-01-17 15:38:28] 55  Archived website id 1, period = week, date = 2022-12-19, segment = '', 1 visits found. Time elapsed: 3.072s
INFO [2023-01-17 15:38:32] 55  Archived website id 1, period = month, date = 2023-01-01, segment = '', 1 visits found. Time elapsed: 2.795s
INFO [2023-01-17 15:38:32] 55  Archived website id 1, period = month, date = 2022-12-01, segment = '', 4 visits found. Time elapsed: 3.557s
INFO [2023-01-17 15:38:35] 55  Archived website id 1, period = year, date = 2023-01-01, segment = '', 1 visits found. Time elapsed: 2.283s
INFO [2023-01-17 15:38:35] 55  Archived website id 1, period = year, date = 2022-01-01, segment = '', 4 visits found. Time elapsed: 2.539s
INFO [2023-01-17 15:38:35] 55  Finished archiving for site 1, 10 API requests, Time elapsed: 13.155s [1 / 3 done]
INFO [2023-01-17 15:38:35] 55  Start processing archives for site 2.
INFO [2023-01-17 15:38:35] 55  Finished archiving for site 2, 0 API requests, Time elapsed: 0.018s [2 / 3 done]
INFO [2023-01-17 15:38:35] 55  Start processing archives for site 3.
INFO [2023-01-17 15:38:35] 55  Finished archiving for site 3, 0 API requests, Time elapsed: 0.013s [3 / 3 done]
INFO [2023-01-17 15:38:35] 55  Done archiving!
INFO [2023-01-17 15:38:35] 55  ---------------------------
INFO [2023-01-17 15:38:35] 55  SUMMARY
INFO [2023-01-17 15:38:35] 55  Processed 10 archives.
INFO [2023-01-17 15:38:35] 55  Total API requests: 10
INFO [2023-01-17 15:38:35] 55  done: 10 req, 13248 ms, no error
INFO [2023-01-17 15:38:35] 55  Time elapsed: 13.248s
INFO [2023-01-17 15:38:35] 55  ---------------------------
INFO [2023-01-17 15:38:35] 55  SCHEDULED TASKS
INFO [2023-01-17 15:38:35] 55  Starting Scheduled tasks...
INFO [2023-01-17 15:38:35] 55  Scheduler: executing task Piwik\Plugins\CustomJsTracker\Tasks.updateTracker...
INFO [2023-01-17 15:38:35] 55  Scheduler: finished. Time elapsed: 0.108s
INFO [2023-01-17 15:38:35] 55  Scheduler: executing task Piwik\Plugins\PrivacyManager\Tasks.deleteLogData...
INFO [2023-01-17 15:38:35] 55  Scheduler: finished. Time elapsed: 0.000s
INFO [2023-01-17 15:38:35] 55  Scheduler: executing task Piwik\Plugins\PrivacyManager\Tasks.anonymizePastData...
INFO [2023-01-17 15:38:35] 55  Scheduler: finished. Time elapsed: 0.002s
INFO [2023-01-17 15:38:35] 55  Scheduler: executing task Piwik\Plugins\GeoIp2\GeoIP2AutoUpdater.update...
INFO [2023-01-17 15:38:35] 55  Downloading https://download.db-ip.com/free/dbip-city-lite-2023-01.mmdb.gz to /var/www/tmp/latest/DBIP-City.mmdb.gz.download.
INFO [2023-01-17 15:38:41] 55  GeoIP2AutoUpdater: successfully downloaded 'https://download.db-ip.com/free/dbip-city-lite-2023-01.mmdb.gz'
INFO [2023-01-17 15:38:42] 55  GeoIP2AutoUpdater: successfully updated GeoIP 2 database 'https://download.db-ip.com/free/dbip-city-lite-2023-01.mmdb.gz'
INFO [2023-01-17 15:38:42] 55  Scheduler: finished. Time elapsed: 6.889s
INFO [2023-01-17 15:38:42] 55  done
INFO [2023-01-17 15:38:42] 55  ---------------------------
```

**Comment by joubertredrat** [comment]
The problem wasn't trailing comma because was implemented in PHP 7.3. https://wiki.php.net/rfc/trailing-comma-function-calls

**Comment by joubertredrat** [comment]
Tested with PHP 7.2.9 and I got the issue, probably their php-fpm version is different from php-cli.

![screenshot-0 0 0 0_50729-2023 01 24-20_18_29](https://user-images.githubusercontent.com/1520407/214442768-850445ff-4a9f-46fc-99d3-0d20b0a09b52.png)
